runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+ *  object via TXBegin(), then dump the frozen object's
+ *  state into CheckpointEntry records into the object's
+ *  stream.
+ *
+ * TODO: Generalize to all SMR objects.
+ */
+public class CheckpointWriter {
+    /** Metadata to be stored in the CP's 'dict' map.
+     */
+    private UUID streamID;
+    private String author;
+    @Getter
+    private UUID checkpointID;
+    private LocalDateTime startTime;
+    private long startAddress, endAddress;
+    private long numEntries = 0, numBytes = 0;
+    Map<String, String> mdKV = new HashMap<>();
+
+    /** Mutator lambda to change map key.  Typically used for
+     *  testing but could also be used for type conversion, etc.
+     */
+    @Getter
+    @Setter
+    Function<Object,Object> keyMutator = (x) -> x;
+
+    /** Mutator lambda to change map value.  Typically used for
+     *  testing but could also be used for type conversion, etc.
+     */
+    @Getter
+    @Setter
+    Function<Object,Object> valueMutator = (x) -> x;
+
+    /** Batch size: number of SMREntry in a single CONTINUATION.
+     */
+    @Getter
+    @Setter
+    private int batchSize = 50;
+
+    /** BiConsumer to run after every CheckpointEntry is appended
+     * to the stream
+     */
+    @Getter
+    @Setter
+    BiConsumer<CheckpointEntry,Long> postAppendFunc = (cp, l) -> {};
+
+    /** Local ref to the object's runtime.
+     */
+    private CorfuRuntime rt;
+
+    /** Local ref to the stream's view.
+     */
+    BackpointerStreamView sv;
+
+    /** Local ref to the object that we're dumping.
+     *  TODO: generalize to all SMR objects.
+     */
+    private SMRMap map;
+
+    public CheckpointWriter(CorfuRuntime rt, UUID streamID, String author, SMRMap map) {
+        this.rt = rt;
+        this.streamID = streamID;
+        this.author = author;
+        this.map = map;
+        checkpointID = UUID.randomUUID();
+        sv = new BackpointerStreamView(rt, streamID);
+    }
+
+    /** Static method for all steps necessary to append checkpoint
+     *  data for an SMRMap into its own stream.
+     */
+
+    public static List<Long> appendCheckpoint(CorfuRuntime rt, UUID streamID, String author, SMRMap map) {
+        return appendCheckpoint(rt, streamID, author, map, cpw -> {});
+    }
+
+    /**
+     *
+     * @param rt Runtime for this map
+     * @param streamID StreamID for this map's underlying stream
+     * @param author String to identify the author of this checkpoint
+     * @param map SMRMap do dump into this checkpoint
+     * @param setupWriter Lambda to transform the new CheckPointWriter
+     *                    object (e.g., change batch size, etc) prior
+     *                    to use.
+     * @return List of global addresses of all entries for this checkpoint.
+     */
+
+    public static List<Long> appendCheckpoint(CorfuRuntime rt, UUID streamID, String author, SMRMap map,
+                                              Consumer<CheckpointWriter> setupWriter) {
```

Comment:
This is static and it uses all the helper methods while still mutating the class members (i.e. mdKV). Can you either remove the mutable class members or make this a nonstatic method ?